analysis/topology/sequences.lean

```diff
@@ -0,0 +1,282 @@
+  exists.intro (λ n, p) ⟨assume n, ‹p∈ M›, const_seq_conv p⟩
+
+def is_seq_closed (A : set X) : Prop := A = sequential_closure A
+
+/- A convenience lemma for showing that a set is sequentially closed. -/
+lemma is_seq_closed_of_def {A : set X} (h : ∀ (x : ℕ → X), (∀ n : ℕ, ((x n) ∈ A)) → ∀ p : X,
+  converges_to x p → p ∈ A) : is_seq_closed A :=
+show A = sequential_closure A, from set.ext (assume p, iff.intro
+(assume : p ∈ A, subset_seq_closure A ‹p ∈ A›)
+(assume :p ∈ sequential_closure A, 
+  have ∃ x : ℕ → X, (∀ n : ℕ, ((x n) ∈ A)) ∧ (converges_to x p), by assumption,
+  let ⟨x, ⟨_, _⟩⟩ := this in
+  show p ∈ A, from h x ‹∀ n : ℕ, ((x n) ∈ A)› p ‹converges_to x p›))
+
+/- The sequential closure of a set is containt in the closure of that set. The converse is not
+ - true.
+ -/ 
+lemma sequential_closure_subset_closure (M : set X) : sequential_closure M ⊆ closure M :=
+suffices ∀ p, p ∈ sequential_closure M → p ∈ closure M, by assumption,
+assume p,
+assume : ∃ x : ℕ → X, (∀ n :ℕ, ((x n) ∈ M)) ∧ (converges_to x p),
+let ⟨x, ⟨_, _⟩⟩ := this in
+show p ∈ closure M, from
+-- we have to show that p is in the closure of M
+-- using mem_closure_iff, this is equivalent to proving that every open neighbourhood
+-- has nonempty intersection with A, but this is witnessed by our sequence x
+suffices ∀ O, is_open O → p ∈ O → O ∩ M ≠ ∅, from mem_closure_iff.mpr this,
+assume O is_open_O O_cap_M_neq_empty,
+let ⟨n0, _⟩ := ‹converges_to x p› O ‹p ∈ O› ‹is_open O› in
+have (x n0) ∈ O, from ‹∀ n ≥ n0, x n ∈ O› n0 (show n0 ≥ n0, from le_refl n0),
+have (x n0) ∈ O ∩ M, from ⟨this, ‹∀n, x n ∈ M› n0⟩,
+set.ne_empty_of_mem this
+
+/- A set is sequentially closed if it is closed. -/
+lemma is_seq_closed_of_is_closed : ∀ M : set X, is_closed M → is_seq_closed M :=
+assume M (_ : is_closed M),
+have M = closure M ,from  eq.symm $ closure_eq_of_is_closed ‹is_closed M›,
+have M ⊆ sequential_closure M, from subset_seq_closure M,
+have sequential_closure M ⊆ M, from 
+  calc sequential_closure M ⊆ closure M : sequential_closure_subset_closure M
+                        ... = M : closure_eq_of_is_closed ‹is_closed M›,
+show M = sequential_closure M, from
+set.eq_of_subset_of_subset ‹M ⊆ sequential_closure M› ‹sequential_closure M ⊆ M›
+
+/- The limit of a convergent sequence in a sequentially closed set is in that set.-/
+lemma is_mem_of_conv_to_of_is_seq_closed {A : set X} (_ : is_seq_closed A) {x : ℕ → X}
+  (_ :∀ n, x n ∈ A) {limit : X} (_ :converges_to x limit) : limit ∈ A :=
+have limit ∈ sequential_closure A, from 
+  show ∃ x : ℕ → X, (∀ n : ℕ, ((x n) ∈ A)) ∧ (converges_to x limit), from
+    exists.intro x ⟨‹∀ n, x n ∈ A›, ‹converges_to x limit›⟩,
+eq.subst (eq.symm ‹is_seq_closed A›) ‹limit ∈ sequential_closure A›
+
+/- The limit of a convergent sequence in a closed set is in that set.-/
+lemma is_mem_of_is_closed_of_conv_to {A : set X} (_ : is_closed A) {x : ℕ → X}
+  (_ :∀ n, x n ∈ A) {limit : X} (_ :converges_to x limit) : limit ∈ A :=
```

Comment:
spaces after colons. This shows up in other places too, I'm not going to comment on all.
```suggestion
  (_ : ∀ n, x n ∈ A) {limit : X} (_ : converges_to x limit) : limit ∈ A :=
```

Reply:
Does that also apply to, say ((1:ℝ)/((n+1):ℝ)) ?

Reply:
In principle yes, but I usually omit them for small casts like that.